### PR TITLE
mgr/dashboard: Adds unit test coverage report

### PR DESCRIPTION
**Pull request by a2batic**
Enable plugin 'Cobertura' to generate code coverage report, for every PR for ceph-dashboard frontend unit test. Cobertura jenkins plugin: https://wiki.jenkins.io/display/JENKINS/Cobertura+Plugin

Fixes: http://tracker.ceph.com/issues/24377
Signed-off-by: Kanika Murarka <kmurarka@redhat.com>

### CODE REVIEW

File: ceph-pull-requests/config/definitions/ceph-pull-requests.yml [hunk 1]
```diff
@@ -62,3 +62,22 @@
                   - ABORTED
               build-steps:
                 - shell: "sudo reboot"
+      - cobertura:
+          report-file: "/src/pybing/mgr/dashboard/frontend/coverage/cobertura-coverage.xml"
```
Comment (joscollin) [inline ceph-pull-requests/config/definitions/ceph-pull-requests.yml]: s/pybing/pybind

**Comment by joscollin** [comment]
s/Unable/Enable ?

File: ceph-pull-requests/config/definitions/ceph-pull-requests.yml [hunk 2]
```diff
@@ -62,3 +62,22 @@
                   - ABORTED
               build-steps:
                 - shell: "sudo reboot"
+      - cobertura:
+          report-file: "src/pybind/mgr/dashboard/frontend/coverage/cobertura-coverage.xml"
+          only-stable: "true"
+          fail-no-reports: "true"
```
Comment (alfredodeza) [inline ceph-pull-requests/config/definitions/ceph-pull-requests.yml]: do you really want to fail the build if no reports are produced?

File: ceph-pull-requests/config/definitions/ceph-pull-requests.yml [hunk 3]
```diff
@@ -62,3 +62,22 @@
...
+          fail-unhealthy: "true"
+          fail-unstable: "true"
+          health-auto-update: "true"
+          stability-auto-update: "true"
+          zoom-coverage-chart: "true"
+          source-encoding: "Big5"
```
Comment (alfredodeza) [inline ceph-pull-requests/config/definitions/ceph-pull-requests.yml]: I don't think there is a need to change from ASCII here, why is `Big5` required?

File: ceph-pull-requests/config/definitions/ceph-pull-requests.yml [hunk 4]
```diff
@@ -62,3 +62,22 @@
                   - ABORTED
               build-steps:
                 - shell: "sudo reboot"
+      - cobertura:
+          report-file: "src/pybind/mgr/dashboard/frontend/coverage/cobertura-coverage.xml"
+          only-stable: "true"
+          fail-no-reports: "true"
+          fail-unhealthy: "true"
```
Comment (alfredodeza) [inline ceph-pull-requests/config/definitions/ceph-pull-requests.yml]: I don't think you want to fail the builds for unhealthy or unstable, these should just be removed (they both default to `false`)

File: ceph-pull-requests/config/definitions/ceph-pull-requests.yml (hunk 2 above)
Comment (a2batic) [reply ceph-pull-requests/config/definitions/ceph-pull-requests.yml]: If build passes, report will be there for sure, so if report is not there, means build fails. So I thought of putting this condition. It should be fine if I remove it as well.

Comment (alfredodeza) [reply ceph-pull-requests/config/definitions/ceph-pull-requests.yml]: Just making sure that is really the intention here. I'm OK if that is the objective

File: ceph-pull-requests/config/definitions/ceph-pull-requests.yml (hunk 3 above)
Comment (a2batic) [reply ceph-pull-requests/config/definitions/ceph-pull-requests.yml]: I checked on few forums, some people mentioned about some garbage characters which were not properly converted to ASCII being showing in report. So, I went with Big5. But, I didnt test will ASCII myself. Thoughts?

Comment (alfredodeza) [reply ceph-pull-requests/config/definitions/ceph-pull-requests.yml]: Sounds reasonable. We can always update this if it is needed later. Thanks for the explanation!

File: ceph-pull-requests/config/definitions/ceph-pull-requests.yml (hunk 4 above)
Comment (a2batic) [reply ceph-pull-requests/config/definitions/ceph-pull-requests.yml]: Yup, `fail-unhealthy` can be removed I think

**Comment by LenzGr** [comment]
Have all requested changes been addressed in the meanwhile? Can this be merged?

**Comment by alfredodeza** [comment]
@LenzGr we can't merge because this needs to be rebased

**Comment by a2batic** [comment]
@joscollin @alfredodeza, rebased the PR. Please review.

**Comment by joscollin** [comment]
<jcollin> dgalloway re https://github.com/ceph/ceph-build/pull/1042: Is Corbetura plugin already installed/available in Jenkins? I hope it is already installed.
<dgalloway> it is now
<jcollin> so you have installed it now?
<dgalloway> yes
<jcollin> Thanks

**Comment by a2batic** [comment]
@joscollin done. Please review.

**Comment by joscollin** [comment]
1. Considering the `target:` values `files: healthy: 10 unhealthy: 20 failing: 30; method: healthy: 50 unhealthy: 40 failing: 30`, Is exercising all options with non-defaults values 
is a good idea? I think these values should be fine-tuned for Ceph, which decides whether there is a failure or not. I was particularly looking at `fail-unhealthy (bool) – Unhealthy projects will be failed (default false)` in the [doc](https://docs.openstack.org/infra/jenkins-job-builder/publishers.html#publishers.cobertura). Will this give a low success rate? @LenzGr  What is your opinion?

2. `fail-no-reports (bool) – fail builds if no coverage reports are found (default false)` - @LenzGr do you think that dependency is necessary?

3. @a2batic Your signed-off (pseudonym) doesn't comply with [Developer's Certificate of Origin 1.1](https://github.com/ceph/ceph/blob/master/SubmittingPatches.rst#developers-certificate-of-origin-11)

4. Missing `coverage` dir in the source tree? Or will it be created automatically?

**Comment by a2batic** [comment]
> Considering the target: values files: healthy: 10 unhealthy: 20 failing: 30; method: healthy: 50 unhealthy: 40 failing: 30, Is exercising all options with non-defaults values
> is a good idea? I think these values should be fine-tuned for Ceph, which decides whether there is a failure or not. I was particularly looking at fail-unhealthy (bool) – Unhealthy projects will be failed (default false) in the doc. Will this give a low success rate? @LenzGr What is your opinion?

@joscollin I am not very sure of threshold values, the one in the the doc seems reasonable, but we can change it later, or maybe we can remove right now and see what type of report is generated.

> @a2batic Your signed-off (pseudonym) doesn't comply with Developer's Certificate of Origin 1.1

Will change this.

> Missing coverage dir in the source tree? Or will it be created automatically?

Yes, it is automatically generated in workspace folder after the build is complete.

**Comment by joscollin** [comment]
Jenkins retest this please

**Comment by joscollin** [comment]
Jenkins retest this please

**Comment by joscollin** [comment]
@a2batic Needs rebase now. https://github.com/ceph/ceph-build/pull/1057 is merged.

**Comment by a2batic** [comment]
@joscollin done

**Pull request merged**